Nodejs/Product/TypeScriptSourceMapReader/SourceMapReader.cs

```diff
@@ -0,0 +1,290 @@
+                if (tsLine < 1) {
+                    // Incorrect start column dont support this map
+                    throw new ErrorDecodingSourcemapException("Invalid tsEntry start line found");
+                }
+
+                // 4. Relative sourceColumn 0 based 
+                tsColumn = tsColumn + Base64VLQFormat.decode(mappingStr, ref i);
+                Debug.Assert(i <= mappingStrLen);
+                if (tsColumn < 1) {
+                    // Incorrect start column dont support this map
+                    throw new ErrorDecodingSourcemapException("Invalid tsEntry start column found");
+                }
+
+                // 5. Check if there is name:
+                int recordNameIndex = -1;
+                if (!IsSourceMappingSegmentEnd(i, mappingStrLen, mappingStr)) {
+                    nameIndex = nameIndex + Base64VLQFormat.decode(mappingStr, ref i);
+                    recordNameIndex = nameIndex;
+                    Debug.Assert(i <= mappingStrLen);
+
+                    if (nameIndex < 0 || nameIndex >= encodedSourceMap.names.Length) {
+                        // Invalid name index found - do not support this map
+                        throw new ErrorDecodingSourcemapException("Invalid name index for the source map entry");
+                    }
+                }
+
+                var jsEntry = new SourceMapSpan(jsLine, jsColumn);
+                var tsEntry = new SourceMapSpan(tsLine, tsColumn);
+
+                var myString = "Emitted (" + jsEntry.StartLine + ", " + jsEntry.StartColumn + ") ";
+                myString = myString + "source (" + tsEntry.StartLine + ", " + tsEntry.StartColumn + ") ";
+                if (recordNameIndex >= 0) {
+                    myString = myString + "name (" + encodedSourceMap.names[recordNameIndex] + ") ";
+                }
+                Debug.WriteLine(myString);
+
+                decodedSourceMap.AddSourceMapSpanMapping(jsEntry, tsEntry, recordNameIndex);
+            }
+
+            return decodedSourceMap;
+        }
+
+        /// <summary>
+        /// Returns if the current index is the end of source code mapping segment end
+        /// </summary>
+        /// <param name="currentIndex">current index in the mappings string</param>
+        /// <param name="mappingsStringLength">length of mappings string</param>
+        /// <param name="mappingsString">mappings string</param>
+        /// <returns>true if current index in the mappingsString indicates end of mapping entry segment</returns>
+        private bool IsSourceMappingSegmentEnd(int currentIndex, int mappingsStringLength, string mappingsString) {
+            if (currentIndex == mappingsStringLength) {
+                return true;
+            }
+
+            if (mappingsString[currentIndex] == ',') {
+                return true;
+            }
+
+            if (mappingsString[currentIndex] == ';') {
+                return true;
+            }
+
+            return false;
```

Comment:
[super-nit] Merge conditional logic to expression: `return currentIndex == mappingsStringLength || ...`